Upgrade TypeScript to version 5

### What version of Remix are you using?

1.14.3

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

Try to upgrade typescript to version 5.x

```
    "typescript": "^5.0.0",
```

> npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: @remix-run/eslint-config@1.14.3
npm ERR! Found: typescript@5.0.2


### Expected Behavior

Remix should be able to work with typescript version 5.  Multiple packages contain the following

```
  "devDependencies": {
    "typescript": "^4.7.4"
  },
  "peerDependencies": {
    "typescript": "^4.0.0"
  },
```

### Actual Behavior

Upgrading typescript is not possible:

```
npm WARN ERESOLVE overriding peer dependency                                                               
npm WARN While resolving: app@1.0.0                  
npm WARN Found: typescript@4.9.5                     
npm WARN node_modules/typescript                     
npm WARN   dev typescript@"^5.0.0" from the root project                                                   
npm WARN   5 more (@remix-run/eslint-config, ...)                                                          
npm WARN                                             
npm WARN Could not resolve dependency:               
npm WARN peerOptional typescript@"^4.0.0" from @remix-run/eslint-config@1.14.0                             
npm WARN node_modules/@remix-run/eslint-config                                                             
npm WARN   dev @remix-run/eslint-config@"^1.7.5" from the root project                  
```

--- a/docs/guides/typescript.md
+++ b/docs/guides/typescript.md
@@ -33,7 +33,7 @@ The Remix compiler will not do any type checking (it simply removes the types). 
     "@types/react": "^17.0.38",
     "@types/react-dom": "^17.0.11",
     "eslint": "^8.23.1",
-    "typescript": "^4.7.4"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

--- a/package.json
+++ b/package.json
@@ -131,7 +131,7 @@
     "tailwindcss": "^3.3.0",
     "to-vfile": "7.2.3",
     "type-fest": "^2.16.0",
-    "typescript": "^4.9.5",
+    "typescript": "^5.0.4",
     "unified": "^10.1.2",
     "unist-util-remove": "^3.1.0",
     "unist-util-visit": "^4.1.1"

--- a/packages/remix-dev/__tests__/fixtures/cloudflare/package.json
+++ b/packages/remix-dev/__tests__/fixtures/cloudflare/package.json
@@ -27,7 +27,7 @@
     "@types/react-dom": "^18.0.8",
     "eslint": "^8.27.0",
     "npm-run-all": "^4.1.5",
-    "typescript": "^4.8.4",
+    "typescript": "^5.0.4",
     "wrangler": "^2.2.1"
   },
   "engines": {

--- a/packages/remix-dev/__tests__/fixtures/indie-stack/package.json
+++ b/packages/remix-dev/__tests__/fixtures/indie-stack/package.json
@@ -16,7 +16,7 @@
     "@types/node": "^17.0.35",
     "@types/react": "^18.0.15",
     "@types/react-dom": "^18.0.6",
-    "typescript": "^4.7.4"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

--- a/packages/remix-dev/__tests__/fixtures/node/package.json
+++ b/packages/remix-dev/__tests__/fixtures/node/package.json
@@ -21,7 +21,7 @@
     "@types/react": "^18.0.25",
     "@types/react-dom": "^18.0.8",
     "eslint": "^8.27.0",
-    "typescript": "^4.8.4"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

--- a/packages/remix-dev/__tests__/fixtures/replace-remix-magic-imports/package.json
+++ b/packages/remix-dev/__tests__/fixtures/replace-remix-magic-imports/package.json
@@ -56,7 +56,7 @@
     "prisma": "^3.11.0",
     "start-server-and-test": "^1.14.0",
     "tailwindcss": "^3.1.0",
-    "typescript": "^4.7.4",
+    "typescript": "^5.0.4",
     "vite-tsconfig-paths": "^3.4.1",
     "vitest": "^0.7.4"
   },

--- a/packages/remix-eslint-config/package.json
+++ b/packages/remix-eslint-config/package.json
@@ -45,7 +45,7 @@
     "eslint": "^8.37.0",
     "jest": "^27.5.1",
     "react": "^18.2.0",
-    "typescript": "^4.9.5"
+    "typescript": "^5.0.4"
   },
   "peerDependencies": {
     "eslint": "^8.0.0",

--- a/packages/remix-testing/package.json
+++ b/packages/remix-testing/package.json
@@ -27,7 +27,7 @@
     "@types/node": "^18.11.9",
     "@types/react": "^18.0.24",
     "@types/react-dom": "^18.0.8",
-    "typescript": "^4.8.4"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

--- a/scripts/playground/template/package.json
+++ b/scripts/playground/template/package.json
@@ -50,7 +50,7 @@
     "tailwindcss": "^3.1.8",
     "ts-node": "^10.9.1",
     "tsconfig-paths": "^4.1.0",
-    "typescript": "^4.8.2"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

--- a/templates/arc/package.json
+++ b/templates/arc/package.json
@@ -26,7 +26,7 @@
     "@types/react-dom": "^18.0.8",
     "eslint": "^8.27.0",
     "npm-run-all": "^4.1.5",
-    "typescript": "^4.8.4"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

--- a/templates/cloudflare-pages/package.json
+++ b/templates/cloudflare-pages/package.json
@@ -27,7 +27,7 @@
     "@types/react-dom": "^18.0.8",
     "eslint": "^8.27.0",
     "npm-run-all": "^4.1.5",
-    "typescript": "^4.8.4",
+    "typescript": "^5.0.4",
     "wrangler": "^2.2.1"
   },
   "engines": {

--- a/templates/cloudflare-workers/package.json
+++ b/templates/cloudflare-workers/package.json
@@ -28,7 +28,7 @@
     "eslint": "^8.27.0",
     "miniflare": "^2.11.0",
     "npm-run-all": "^4.1.5",
-    "typescript": "^4.8.4",
+    "typescript": "^5.0.4",
     "wrangler": "^2.2.1"
   },
   "engines": {

--- a/templates/express/package.json
+++ b/templates/express/package.json
@@ -30,7 +30,7 @@
     "eslint": "^8.27.0",
     "nodemon": "^2.0.20",
     "npm-run-all": "^4.1.5",
-    "typescript": "^4.8.4"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

--- a/templates/fly/package.json
+++ b/templates/fly/package.json
@@ -22,7 +22,7 @@
     "@types/react": "^18.0.25",
     "@types/react-dom": "^18.0.8",
     "eslint": "^8.27.0",
-    "typescript": "^4.8.4"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

--- a/templates/netlify/package.json
+++ b/templates/netlify/package.json
@@ -24,7 +24,7 @@
     "@types/react": "^18.0.25",
     "@types/react-dom": "^18.0.8",
     "eslint": "^8.27.0",
-    "typescript": "^4.8.4"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

--- a/templates/remix/package.json
+++ b/templates/remix/package.json
@@ -21,7 +21,7 @@
     "@types/react": "^18.0.25",
     "@types/react-dom": "^18.0.8",
     "eslint": "^8.27.0",
-    "typescript": "^4.8.4"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

--- a/templates/vercel/package.json
+++ b/templates/vercel/package.json
@@ -22,7 +22,7 @@
     "@types/react": "^18.0.25",
     "@types/react-dom": "^18.0.8",
     "eslint": "^8.27.0",
-    "typescript": "^4.8.4"
+    "typescript": "^5.0.4"
   },
   "engines": {
     "node": ">=14"

